### DISCUSSION TITLE
String encoding

### DISCUSSION THREAD

**Discussion by aelhadee** [named-data/python-ndn#61] [General]
Hello, 
I noticed when sending strings over NDN, it takes longer, so are there any additional steps besides encoding (assuming utf-8) and packing the encoded data in TLV packets? It was compared against JSON w/ utf-8 encoded strings in a different networking protocol and NDN was taking a little longer to transmit the data.

**Comment by zjkmxy** [comment]
Hello. I'm sorry but I don't understand what do you mean by "it takes longer". Could you please explain how you did the experiment? Thanks

**Reply by aelhadee** [reply]
I am calculating the difference of each received packet at the receiver. For example, I send 1248*2 strings every 1 ms using the same setup for NDN compared with RTPS. 

In the case of 1248*2 strings, NDN had a mean of 4.5 ms and 3.75 ms when sending over TCP and UDP compared to 2.44 ms when RTPS is used, so I assumed that is due to the encoding time and packing the encoded data in TLV and wanted to confirm.

**Reply by zjkmxy** [reply]
Hello, there could be several possible reasons:
- The NDN forwarder (NFD) is not optimized for performance.
- Your NDN connection between two nodes could be over TCP/UDP, which means it can never overperform TCP.
- python-ndn has all encoding code written in Python, which makes it slower than other libraries (such as ndn-cxx and go-ndn).

I don't think millisecond level performance difference is important in an application scenario, as the propagation delay is already 10~100ms level. But if you really need this level of performance, I think you may want to try ndn-cxx.
